Updating team model

rchan26 opened the pull request:
Working towards #210 and hopefully completing #141.

Updating team model to use features that we added to the bpl model to predict World Cup games (https://github.com/alan-turing-institute/WorldCupPrediction)

Working to update the bpl interface to work with `NeutralDixonColesMatchPredictor` model (see https://github.com/anguswilliams91/bpl-next/tree/feature/neutral-dixon-coles branch). This required a few other updates to functions / pipelines that used the team model to be able to use this model and importantly pass in a choice for a hyperparamter `epsilon` which controls how much to downweight games from the past in the likelihood.

While working on this, I also added some type-hints to some functions that the team model used in `airsenal.framework.utils` and `airsenal.framework.prediction_utils`.

Example:
```
weight = 0.3
# pass epsilon as a keyword argument
model_team = get_fitted_team_model(
    season="2122",
    gameweek=1,
    session=session,
    model=NeutralDixonColesMatchPredictor(),
    epsilon=weight
)
```

If no time-weighting is desired, simply do not pass anything to epsilon.

With #591, this should also work with ExtendedDixonColesMatchPredictor too:
```
weight = 0.3
# pass epsilon as a keyword argument
model_team = get_fitted_team_model(
    season="2122",
    gameweek=1,
    session=session,
    model=ExtendedDixonColesMatchPredictor(),
    epsilon=weight
)
```

Can use CLI to fit model with time weighting, e.g. to use extended model with `epsilon=5.0`:
```
airsenal_run_prediction --epsilon 5
```

Also, to use neutral model with 5 week ahead look, by default `epsilon=0.0` if none is passed:
```
airsenal_run_prediction --weeks_ahead 5 --team_model "neutral"
```

Also can pass these to `airsenal_run_pipeline` too.

rchan26 commented:
currently failing tests when encounter a `NaT` time in trying to obtain time difference between fixtures and the current gameweek

rchan26 commented:
tests pass locally, but fail currently as PR to `bpl-next` has not been accepted just yet: https://github.com/anguswilliams91/bpl-next/pull/19.

rchan26 commented:
to optimise the team model (rather than using `replay_season` functionality to optimise over the full AIrsenal model):
- compare performance of extended and neutral model (with different values of epsilon) on predicting the number of goals scored by each team
- metrics to consider: brier, rps, cross-entropy

rchan26 commented:
@jack89roberts - I think this functionally makes us able to use the time weighting of extended and neutral models. Defaults to use extended with no time weighting though!

jack89roberts commented on airsenal/framework/bpl_interface.py:
```diff
@@ -63,70 +85,150 @@ def get_ratings_dict(season, teams, dbsession):
-):
+    gameweek: int,
+    season: str = CURRENT_SEASON,
+    team_model: Optional[
+        Union[ExtendedDixonColesMatchPredictor, NeutralDixonColesMatchPredictor]
+    ] = None,
+    dbsession: Session = session,
+    ratings: bool = True,
+    model: Union[
+        ExtendedDixonColesMatchPredictor, NeutralDixonColesMatchPredictor
+    ] = ExtendedDixonColesMatchPredictor(),
```
Can we avoid this function taking both a `team_model` and `model` argument? It's a bit confusing.

rchan26 replied:
yes sure, I agree this is confusing

I've removed one of them and now we fit if either no model is passed, or if the model is not fitted yet (which will be the case if model.teams is None)

jack89roberts commented:
Happy to merge this to develop now, unless there's anything else you prefer to do on this branch first.

jack89roberts commented:
Actually, I get errors from `airsenal_run_pipeline`:

```
Fitting team model (<class 'str'>)...
Fitting <class 'str'> model with epsilon = 0.0
Traceback (most recent call last):
  File "/Users/jroberts/micromamba/envs/airsenalenv/bin/airsenal_run_pipeline", line 8, in <module>
    sys.exit(run_pipeline())
  File "/Users/jroberts/micromamba/envs/airsenalenv/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jroberts/micromamba/envs/airsenalenv/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/jroberts/micromamba/envs/airsenalenv/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jroberts/micromamba/envs/airsenalenv/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/scripts/airsenal_run_pipeline.py", line 173, in run_pipeline
    predict_ok = run_prediction(
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/scripts/airsenal_run_pipeline.py", line 265, in run_prediction
    tag = make_predictedscore_table(
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/scripts/fill_predictedscore_table.py", line 206, in make_predictedscore_table
    calc_all_predicted_points(
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/scripts/fill_predictedscore_table.py", line 102, in calc_all_predicted_points
    model_team = get_fitted_team_model(
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/framework/bpl_interface.py", line 176, in get_fitted_team_model
    team_model = create_and_fit_team_model(
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/framework/bpl_interface.py", line 128, in create_and_fit_team_model
    return model.fit(training_data=training_data, **fit_args)
AttributeError: 'str' object has no attribute 'fit'
```

and `airsenal_replay_season --season 2223 --gameweek_start 1 --gameweek_end 4 --weeks_ahead 2 --num_thread 4`:

```
Traceback (most recent call last):
  File "/Users/jroberts/micromamba/envs/airsenalenv/bin/airsenal_replay_season", line 8, in <module>
    sys.exit(main())
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/scripts/replay_season.py", line 243, in main
    replay_season(
  File "/Users/jroberts/GitHub/AIrsenal/airsenal/scripts/replay_season.py", line 102, in replay_season
    tag = make_predictedscore_table(
TypeError: make_predictedscore_table() got an unexpected keyword argument 'team_model_class'
```

because there are functions passing strings where a model instance is expected (pipeline) or variables names have changed (replay).

rchan26 commented:
ah that's strange, i must've introduced those last night. will fix

rchan26 commented:
hi Jack, I think this will fix it, but I haven't tried to run those commands again. to fix this, I've just wrote this function `airsenal.framework.utils.parse_team_model_from_str` which returns the model instance from the string, and these are called in `airsenal_run_pipeline` and replay season functionality (I think!)

rchan26 commented:
just ran `airsenal_run_pipeline` and `airsenal_replay_season --season 2223 --gameweek_start 1 --gameweek_end 4 --weeks_ahead 2 --num_thread 4` and both seemed to work for me. think this is ready again @jack89roberts - sorry for the delay!

jack89roberts commented on airsenal/framework/bpl_interface.py:
```diff
@@ -33,17 +38,34 @@ def get_result_dict(season, gameweek, dbsession):
+    )
+    end_date = pd.to_datetime(
+        [f.date for f in get_fixtures_for_gameweek(gameweek, season, dbsession)]
+    ).min()
```
Not played with it but should this be `.max()`? I think `time_diff` is meant to be positive so wouldn't we want everything relative to the last fixture in the gameweek?

rchan26 replied:
So I put `.min()` here because don't we want the time difference between the games to the _first_ game in the gameweek? If we had `max()`, the difference would be the latest game in the gameweek (I _think_).

jack89roberts replied:
Ah, we don't use the `time_diff` for making predictions, only for fitting the model, right? That's what I was missing yesterday (I was getting confused about time differences to future matches). So yes `.min()` is right in that case 👍.

jack89roberts commented:
Added a couple of comments above but this looks good to me, thanks @rchan26 ! I'll just try running it myself before merging (probably tomorrow).

jack89roberts commented:
Works for me so just the `.min()` vs. `.max()` comment for the time diff above to think through before merging.

jack89roberts commented:
also @rchan26 I get different results from the optimisation on this vs. develop. Does this change the default model to be one with non-zero time weighting?

rchan26 commented on airsenal/framework/bpl_interface.py:
```diff
@@ -63,70 +85,162 @@ def get_ratings_dict(season, teams, dbsession):
+        print(f"Fitting {type(model)} model with epsilon = {fit_args['epsilon']}")
+    else:
+        print(
+            f"Fitting {type(model)} model but no epsilon passed, so setting epsilon = 0"
```
yes, so I think having `epsilon=0` means that you raise things to power 0 and it should be like having no time-weighting at all

jack89roberts replied:
minor but maybe the message could be tweaked to "using the default" rather than "setting", since the value of epsilon isn't set here

rchan26 replied:
just changed this

rchan26 commented:
@jack89roberts thanks for checking that - we should be careful on that. Is it quite drastically different to what it is on `develop`? I imagine there will always be some randomness but if it's very different then it's worth figuring out why

I thought the default would be to set the same model, but I can double check this

jack89roberts commented:
From a quick look it's not massively different but more different than random I think (up to around 1pt difference in predicted points over 7 weeks, and enough to change the recommended transfers), see below.

### Points prediction

This branch:

```
==================================================
PREDICTED TOP 5 PLAYERS FOR GAMEWEEK(S) [10, 11, 12, 13, 14, 15, 16]:
==================================================
GK:
1. Alisson Ramses Becker, 29.94pts (£5.5m, LIV)
2. Sam Johnstone, 29.14pts (£4.5m, CRY)
3. José Malheiro de Sá, 28.03pts (£5.0m, WOL)
4. Alphonse Areola, 26.74pts (£4.2m, WHU)
5. Guglielmo Vicario, 26.57pts (£5.2m, TOT)
-------------------------
DEF:
1. Kieran Trippier, 30.72pts (£7.0m, NEW)
2. Pedro Porro, 29.23pts (£5.2m, TOT)
3. Vladimír Coufal, 28.78pts (£4.6m, WHU)
4. William Saliba, 28.78pts (£5.2m, ARS)
5. Joel Matip, 27.60pts (£4.9m, LIV)
-------------------------
MID:
1. Mohamed Salah, 50.41pts (£12.8m, LIV)
2. Son Heung-min, 37.77pts (£9.5m, TOT)
3. Phil Foden, 36.54pts (£7.6m, MCI)
4. James Maddison, 36.44pts (£8.1m, TOT)
5. Bruno Borges Fernandes, 36.40pts (£8.3m, MUN)
-------------------------
FWD:
1. Erling Haaland, 53.96pts (£14.0m, MCI)
2. Julián Álvarez, 38.63pts (£7.1m, MCI)
3. Ollie Watkins, 32.99pts (£8.3m, AVL)
4. Yoane Wissa, 29.85pts (£5.9m, BRE)
5. Carlton Morris, 28.01pts (£5.5m, LUT)
-------------------------
```

Develop branch:

```
==================================================
PREDICTED TOP 5 PLAYERS FOR GAMEWEEK(S) [10, 11, 12, 13, 14, 15, 16]:
==================================================
GK:
1. Alisson Ramses Becker, 30.66pts (£5.5m, LIV)
2. Sam Johnstone, 29.37pts (£4.5m, CRY)
3. José Malheiro de Sá, 27.71pts (£5.0m, WOL)
4. Guglielmo Vicario, 27.03pts (£5.2m, TOT)
5. Alphonse Areola, 27.01pts (£4.2m, WHU)
-------------------------
DEF:
1. Kieran Trippier, 30.56pts (£7.0m, NEW)
2. Pedro Porro, 29.85pts (£5.2m, TOT)
3. William Saliba, 29.65pts (£5.2m, ARS)
4. Vladimír Coufal, 29.17pts (£4.6m, WHU)
5. Joel Matip, 28.43pts (£4.9m, LIV)
-------------------------
MID:
1. Mohamed Salah, 51.93pts (£12.8m, LIV)
2. Bruno Borges Fernandes, 38.69pts (£8.3m, MUN)
3. Son Heung-min, 38.21pts (£9.5m, TOT)
4. Marcus Rashford, 37.32pts (£8.7m, MUN)
5. Phil Foden, 36.93pts (£7.6m, MCI)
-------------------------
FWD:
1. Erling Haaland, 54.49pts (£14.0m, MCI)
2. Julián Álvarez, 38.92pts (£7.1m, MCI)
3. Ollie Watkins, 33.31pts (£8.3m, AVL)
4. Yoane Wissa, 28.62pts (£5.9m, BRE)
5. Michail Antonio, 27.78pts (£5.8m, WHU)
-------------------------
```

### Optimisation

Only copied in the total scores and suggested transfer for the first week.

This branch:

```
Strategy for Team ID: 1822891
Baseline score: 369.2424278755185
Best score: 374.1335385574875
 ===============================================
 ========= Optimum strategy ====================
 ===============================================

 =========== Gameweek 10 ================

Chips played:  None

Players in:			Players out:
-----------			------------
Joel Matip			Kyle Walker
```

Develop branch:

```
Baseline score: 375.22877444744825
Best score: 382.5982145613103
 ===============================================
 ========= Optimum strategy ====================
 ===============================================

 =========== Gameweek 10 ================

Chips played:  None

Players in:			Players out:
-----------			------------
Pau Torres			Ethan Pinnock
```

These are just the first transfers from a 7 week optimisation (+ the overall scores). Would be better to compare the predicted points

rchan26 commented:
so I think I may have identified a reason why we might see differences - I changed the default to _not_ use ratings (used to have `ratings=True` by default but I changed this to False in getting the training data. Also for adding new teams, I added and option to set the ratings and this was also defaulted to False. I can change these back to True and maybe that answers why there's differences

rchan26 commented:
same for 7 weeks:

## Points prediction

on this branch:
```
==================================================
PREDICTED TOP 5 PLAYERS FOR GAMEWEEK(S) [11, 12, 13, 14, 15, 16, 17]:
==================================================
GK:
1. Alisson Ramses Becker, 29.86pts (£5.6m, LIV)
2. Sam Johnstone, 28.33pts (£4.5m, CRY)
3. Guglielmo Vicario, 27.88pts (£5.2m, TOT)
4. Alphonse Areola, 27.34pts (£4.3m, WHU)
5. José Malheiro de Sá, 27.30pts (£5.0m, WOL)
-------------------------
DEF:
1. Kieran Trippier, 30.55pts (£7.0m, NEW)
2. Pedro Porro, 29.74pts (£5.2m, TOT)
3. Vladimír Coufal, 29.41pts (£4.6m, WHU)
4. William Saliba, 28.63pts (£5.2m, ARS)
5. Kyle Walker, 27.77pts (£5.3m, MCI)
-------------------------
MID:
1. Mohamed Salah, 50.72pts (£12.8m, LIV)
2. Phil Foden, 43.31pts (£7.5m, MCI)
3. Son Heung-min, 38.77pts (£9.6m, TOT)
4. Bruno Borges Fernandes, 38.53pts (£8.3m, MUN)
5. James Maddison, 37.67pts (£8.1m, TOT)
-------------------------
FWD:
1. Erling Haaland, 57.64pts (£14.0m, MCI)
2. Julián Álvarez, 38.92pts (£7.2m, MCI)
3. Ollie Watkins, 31.79pts (£8.4m, AVL)
4. Eddie Nketiah, 30.60pts (£5.5m, ARS)
5. Yoane Wissa, 28.71pts (£5.9m, BRE)
-------------------------
```

on develop:
```
==================================================
PREDICTED TOP 5 PLAYERS FOR GAMEWEEK(S) [11, 12, 13, 14, 15, 16, 17]:
==================================================
GK:
1. Alisson Ramses Becker, 29.86pts (£5.6m, LIV)
2. Sam Johnstone, 28.33pts (£4.5m, CRY)
3. Guglielmo Vicario, 27.88pts (£5.2m, TOT)
4. Alphonse Areola, 27.34pts (£4.3m, WHU)
5. José Malheiro de Sá, 27.30pts (£5.0m, WOL)
-------------------------
DEF:
1. Kieran Trippier, 30.55pts (£7.0m, NEW)
2. Pedro Porro, 29.74pts (£5.2m, TOT)
3. Vladimír Coufal, 29.41pts (£4.6m, WHU)
4. William Saliba, 28.63pts (£5.2m, ARS)
5. Kyle Walker, 27.77pts (£5.3m, MCI)
-------------------------
MID:
1. Mohamed Salah, 50.72pts (£12.8m, LIV)
2. Phil Foden, 43.31pts (£7.5m, MCI)
3. Son Heung-min, 38.77pts (£9.6m, TOT)
4. Bruno Borges Fernandes, 38.53pts (£8.3m, MUN)
5. James Maddison, 37.67pts (£8.1m, TOT)
-------------------------
FWD:
1. Erling Haaland, 57.64pts (£14.0m, MCI)
2. Julián Álvarez, 38.92pts (£7.2m, MCI)
3. Ollie Watkins, 31.79pts (£8.4m, AVL)
4. Eddie Nketiah, 30.60pts (£5.5m, ARS)
5. Yoane Wissa, 28.71pts (£5.9m, BRE)
```

## Optimisation

This branch:
```
 ===============================================
 ========= Optimum strategy ====================
 ===============================================

 =========== Gameweek 11 ================

Chips played:  None

Players in:			Players out:
-----------			------------
Phil Foden			Elliot Anderson
Eddie Nketiah			Gabriel Fernando de Jesus

...

=== starting 11 ===


== GK ==

André Onana (MUN)

== DEF ==

Joško Gvardiol (MCI)
Levi Colwill (CHE)
Matty Cash (AVL)

== MID ==

Bukayo Saka (ARS)
Mohamed Salah (LIV)(VC)
Moussa Diaby (AVL)
Phil Foden (MCI)

== FWD ==

Erling Haaland (MCI)(C)
Ollie Watkins (AVL)
Eddie Nketiah (ARS)

=== subs ===

Matt Turner (NFO)
Cole Palmer (CHE)
Dan Burn (NEW)
Malo Gusto (CHE)
```

On develop:
```
Baseline score: 326.23015374222854
Best score: 370.29137171826363
 ===============================================
 ========= Optimum strategy ====================
 ===============================================

 =========== Gameweek 11 ================

Chips played:  None

Players in:			Players out:
-----------			------------
Phil Foden			Elliot Anderson
Eddie Nketiah			Gabriel Fernando de Jesus

....

=== starting 11 ===


== GK ==

André Onana (MUN)

== DEF ==

Joško Gvardiol (MCI)
Levi Colwill (CHE)
Matty Cash (AVL)

== MID ==

Bukayo Saka (ARS)
Mohamed Salah (LIV)(VC)
Moussa Diaby (AVL)
Phil Foden (MCI)

== FWD ==

Erling Haaland (MCI)(C)
Ollie Watkins (AVL)
Eddie Nketiah (ARS)

=== subs ===

Matt Turner (NFO)
Cole Palmer (CHE)
Dan Burn (NEW)
Malo Gusto (CHE)
```

jack89roberts commented:
LGTM, thanks!
